worlds/blasphemous/Rules.py

```diff
@@ -1,4458 +1,1373 @@
-
-
-def opened_wothp_gate(state: CollectionState, player: int) -> bool:
-    return state.has_any({"D09Z01S05[W]", "D09Z01S05[SE]", "D09Z01S05[NE]"}, player)
-
-
-def opened_botss_ladder(state: CollectionState, player: int) -> bool:
-    return state.has_any({"D17Z01S04[N]", "D17Z01S04[FrontR]"}, player)
-
-
-def upwarp_skips_allowed(logic: int) -> bool:
-    return logic >= 2
-
-
-def mourning_skips_allowed(logic: int) -> bool:
-    return logic >= 2
-
-
-def enemy_skips_allowed(logic: int, enemy: int) -> bool:
-    return logic >= 2 and enemy == 0
-
-
-def obscure_skips_allowed(logic):
-    return logic >= 2
-
-
-def precise_skips_allowed(logic):
-    return logic >= 2
-
-
-def can_beat_boss(state: CollectionState, boss: str, logic: int, player: int) -> bool:
-    def has_boss_strength(name: str) -> bool:
-        silver: int = state.count("Quicksilver", player) if state.has("D01Z05S27[E]", player) else 0
-        flasks: int = state.count("Empty Bile Flask", player) if \
-            state.has_any({"D01Z05S18[E]", "D02Z02S09[E]", "D03Z02S14[E]", "D03Z03S03[SE]", "D04Z02S13[W]", \
-                "D05Z01S12[E]", "D20Z01S08[W]"}, player) else 0
-        
-
-        playerStrength: float = state.count("Life Upgrade", player) * 0.25 / 6 + \
-            state.count("Mea Culpa Upgrade", player) * 0.25 / 7 + state.count("Fervour Upgrade", player) * 0.20 \
-                / 6 + flasks * 0.15 / 8 + silver * 0.15 / 5
-        
-        bosses: Dict[str, int] = {
-            "warden": -0.10,
-            "ten-piedad": 0.05,
-            "charred-visage": 0.20,
-            "tres-angustias": 0.15,
-            "esdras": 0.25,
-            "melquiades": 0.25,
-            "exposito": 0.30,
-            "quirce": 0.35,
-            "crisanta": 0.50,
-            "isidora": 0.70,
-            "sierpes": 0.70,
-            "amanecida": 0.60,
-            "laudes": 0.60,
-            "perpetua": -0.05,
-            "legionary": 0.20
-        }
-
-        bossStrength: int = bosses[name]
-
-        return playerStrength >= (bossStrength - 0.10 if logic >= 2 else (bossStrength if logic >= 1 else bossStrength + 0.10))
     
-    if boss == "Brotherhood":
+    def opened_dc_ladder(self, state: CollectionState) -> bool:
         return (
-            has_boss_strength("warden")
-            and state.has_any({"D17Z01S11[W]", "D17Z01S11[E]"}, player)
+            state.can_reach_region("D01Z05S25[NE]", self.player)
+            or state.can_reach_region("D01Z05S02[S]", self.player)
         )
-    elif boss == "Mercy":
+    
+    def opened_wotw_cave(self, state: CollectionState) -> bool:
         return (
-            has_boss_strength("ten-piedad")
-            and state.has_any({"D01Z04S18[W]", "D01Z04S18[E]"}, player)
+            state.can_reach_region("D02Z01S01[SW]", self.player)
+            or self.wall_climb(state)
+            and state.can_reach_region("D02Z01S08[E]", self.player)
+            or state.can_reach_region("D02Z01S02[]", self.player)
```

Comment:
I assume `D02Z01S02[]` is correct? It's used twice, but it strikes me as odd that it has no directional indicator

Reply:
Yes, it looks weird but this is correct